### PR TITLE
Direct Link: Dedicated VLAN Mapping

### PR DESCRIPTION
**Pull request by MalarvizhiK**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates OR Closes #0000

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccXXX'

...
```

### CODE REVIEW

**Comment by hkantare** [comment]
@MalarvizhiK 
Can you fix the checks

**Comment by MalarvizhiK** [comment]
The GA for the feature is scheduled in 1Q 2024. Hence, I will wait some more time.

**Comment by MalarvizhiK** [comment]
@hkantare As mentioned earlier, VLAN tagging enablement for DL dedicated feature is GAed. 

We want to publish the terraform code for the same.

**Comment by MalarvizhiK** [comment]
Output from acceptance testing:
```
malark@Malars-MacBook-Pro terraform-provider-ibm %  make testacc TEST=./ibm/service/directlink TESTARGS='-run=TestAccIBMDLGatewayVC_basic'     
==> Checking that code complies with gofmt requirements...

....
=== RUN   TestAccIBMDLGatewayVC_basic
Create Gateway ....0 VLAN Value Create 0 VLAN Value Create --- PASS: TestAccIBMDLGatewayVC_basic (131.31s)
PASS
ok  	github.com/IBM-Cloud/terraform-provider-ibm/ibm/service/directlink	133.202s
```

**Comment by MalarvizhiK** [comment]
Output from acceptance testing
```
malark@Malars-MacBook-Pro terraform-provider-ibm %  make testacc TEST=./ibm/service/directlink TESTARGS='-run=TestAccIBMDLGatewayConnect_basic'
==> Checking that code complies with gofmt requirements...

=== RUN   TestAccIBMDLGatewayConnect_basic
Create Gateway ....--- PASS: TestAccIBMDLGatewayConnect_basic (649.04s)
PASS
ok  	github.com/IBM-Cloud/terraform-provider-ibm/ibm/service/directlink	650.940s
```

File: ibm/service/directlink/resource_ibm_dl_gateway.go [hunk 1]
```diff
@@ -859,6 +879,16 @@ func resourceIBMdlGatewayCreate(d *schema.ResourceData, meta interface{}) error
...
 		}
+		vlan := d.Get(dlVlan)
+		fmt.Printf("%v VLAN Value Create ", vlan)
+
+		if vlan, ok := d.GetOk(dlVlan); ok {
+			mapped_vlan := int64(vlan.(int))
+			gatewayDedicatedTemplateModel.Vlan = &mapped_vlan
+		} else {
+			vlan := d.Get(dlVlan)
+			fmt.Printf("%v VLAN Value Create ", vlan)
```
Comment (hkantare) [inline ibm/service/directlink/resource_ibm_dl_gateway.go:890]: Remove this fmt stmt

File: ibm/service/directlink/resource_ibm_dl_gateway.go [hunk 2]
```diff
@@ -1070,7 +1100,8 @@ func resourceIBMdlGatewayRead(d *schema.ResourceData, meta interface{}) error {
 	}
 	log.Printf("[INFO] Calling getgateway api: %s", dtype)
 
-	instance, response, err := directLink.GetGateway(getOptions)
+	instanceIntf, response, err := directLink.GetGateway(getOptions)
+	instance := instanceIntf.(*directlinkv1.GetGatewayResponse)
```
Comment (hkantare) [inline ibm/service/directlink/resource_ibm_dl_gateway.go:1104]: Can we add some nil check before instanceIntf.(*directlinkv1.GetGatewayResponse) to suppress any nil pinter execeptions

File: ibm/service/directlink/resource_ibm_dl_gateway.go [hunk 3]
```diff
@@ -451,9 +452,19 @@ func ResourceIBMDLGateway() *schema.Resource {
...
-				Computed:    true,
-				Description: "VLAN allocated for this gateway",
+				Type:          schema.TypeInt,
+				Optional:      true,
+				Computed:      true,
+				Description:   "VLAN allocated for this gateway",
+				ConflictsWith: []string{"remove_vlan"},
+				ValidateFunc:  validate.InvokeValidator("ibm_dl_gateway", dlVlan),
+			},
+			dlRemoveVlan: {
```
Comment (hkantare) [inline ibm/service/directlink/resource_ibm_dl_gateway.go:462]: remove_vlan comes only during update scenario?

**Comment by hkantare** [comment]
Share test execuitpn results of respective res/ds

File: ibm/service/directlink/resource_ibm_dl_gateway.go (hunk 1 above)
Comment (MalarvizhiK) [reply ibm/service/directlink/resource_ibm_dl_gateway.go:890]: Done

File: ibm/service/directlink/resource_ibm_dl_gateway.go (hunk 2 above)
Comment (MalarvizhiK) [reply ibm/service/directlink/resource_ibm_dl_gateway.go:1104]: Done

**Comment by MalarvizhiK** [comment]
Data block results

```
data "ibm_dl_gateway" "test_dl_gateway" {
                        name = ibm_dl_gateway.test_dl_gateway_new.name
                 }

output "dl_instances" {
  value = data.ibm_dl_gateway.test_dl_gateway
}

Outputs:


dl_instances = {
    as_prepends                 = []
    bgp_asn                     = 64999
    bgp_cer_cidr                = "169.254.0.22/30"
    bgp_ibm_asn                 = 13884
    bgp_ibm_cidr                = "169.254.0.21/30"
    bgp_status                  = "connect"
    bgp_status_updated_at       = "2024-03-13T07:00:38.738Z"
    connection_mode             = "direct"
    created_at                  = "2024-03-13T06:46:46.475Z"
    crn                         = "crn:v1:staging:public:directlink:dal09:a/3b1bd7fa2bc3406ea70ba4ade8aa3f1b::dedicated:35e37215-e349-459e-a45f-82eb4adfadf1"
    cross_connect_router        = "LAB-xcr01.dal09"
    default_export_route_filter = "permit"
    default_import_route_filter = "permit"
    gateway_vcs                 = []
    global                      = true
    id                          = "35e37215-e349-459e-a45f-82eb4adfadf1"
    link_status                 = "down"
    link_status_updated_at      = "2024-03-13T06:46:43.074Z"
    location_display_name       = "Dallas 09"
    location_name               = "dal09"
    metered                     = true
    name                        = "gatewaynew"
    operational_status          = "awaiting_loa"
    resource_group              = "bf823d4f45b64ceaa4671bee0479346e"
    speed_mbps                  = 1000
    type                        = "dedicated"
    vlan                        = 3967
}



```

**Pull request merged**